### PR TITLE
(VDB-1249) backfill storage

### PR DESCRIPTION
**Pull request by rmulhol**
TODO:
- [x] inline getStorageValue command to backfillStorage
- [x] intelligent insert backfill diff function (no dupes)

### CODE REVIEW

**Comment by rmulhol** [comment]
Dockerfile updates will be made in transformers repo since config is required

File: db/migrations/20200331113100_create_create_back_filled_diff_function.sql [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    empty_storage_value BYTEA := (
+        SELECT '\x0000000000000000000000000000000000000000000000000000000000000000'
+        );
+    inserted_diff_id   INTEGER;
+BEGIN
+    IF last_storage_value = create_back_filled_diff.storage_value THEN
+        RETURN 0;
+    END IF;
+
+    IF last_storage_value is null and create_back_filled_diff.storage_value = empty_storage_value THEN
```
[the remark this answers is not in the document]
Comment (elizabethengelman) [reply db/migrations/20200331113100_create_create_back_filled_diff_function.sql:25]: Oh, I'm seeing now that `HeaderRepository. GetHeadersInRange` sorts the headers it's returning, nice!

File: cmd/backfillStorage.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+var backfillStorageCmd = &cobra.Command{
+	Use:   "backfillStorage",
+	Short: "Backfill smart contract storage for a range of blocks",
+	Long: `Fetch and persist contract storage for a range of blocks. Useful if you have started watching diffs from a
+new contract but do not have storage data from before you started running the transformer.
+Requires a config file structured the same as it would be for running compose or composeAndExecute (to specify which
+addresses and storage slots must be back-filled).
+Requires CLI flags are backfill-storage-start-block (-s) and backfill-storage-end-block (-e) to define range of blocks
+that need to be back-filled.
+Optional CLI flag is backfill-storage-contract-address (-a) to specify a single contract address that needs to be
```
[the remark this answers is not in the document]
Comment (rmulhol) [reply cmd/backfillStorage.go:33]: Yeah bc still need the transformer in the exporter for the keys loader

File: db/migrations/20200331113100_create_create_back_filled_diff_function.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (rmulhol) [reply db/migrations/20200331113100_create_create_back_filled_diff_function.sql:25]: Yeah, still a good thought - will think about ways we can verify that all headers have been synced/checked. Seems like we could verify that we have all headers in the range and that they have a check count > 0, at least

File: libraries/shared/storage/diff_repository.go [hunk 3]
```diff
@@ -55,6 +55,14 @@ func (repository diffRepository) CreateStorageDiff(rawDiff types.RawDiff) (int64
 	return storageDiffID, err
 }
 
+func (repository diffRepository) CreateBackFilledStorageValue(rawDiff types.RawDiff) (int64, error) {
```
[the remark this answers is not in the document]
Comment (rmulhol) [reply libraries/shared/storage/diff_repository.go]: Dig it, removed 👍

**Comment by rmulhol** [comment]
This is pretty slow running on my local machine (~1:30/block, with an incomplete set of keys). Will continue brainstorming ways to speed it up. Unfortunately only getting relevant keys (e.g. exclude entries in mappings and dynamic arrays where event was fired at a later block) imposes some additional cost since we'd need to refresh keys for every header (as opposed to once for all headers), but maybe that's worth it if we can exclude rpc calls/db inserts 🤔

**Comment by rmulhol** [comment]
Update: batching improves performance by 80% locally to 18s/block 🎉

File: pkg/eth/blockchain.go [hunk 4]
```diff
@@ -104,6 +104,33 @@ func (blockChain *BlockChain) LastBlock() (*big.Int, error) {
 	return block.Number, err
 }
 
+func (blockChain *BlockChain) BatchGetStorageAt(account common.Address, keys []common.Hash, blockNumber *big.Int) (map[common.Hash][]byte, error) {
```
Comment (elizabethengelman) [inline pkg/eth/blockchain.go:107]: 🎉

**Pull request merged**